---
title: "Ćwiczenia 8 KF"
output: html_document
date: "2024-12-03"
---

#Zadanie1
library(VariantTools)
library(Rsamtools)
library(GenomicRanges)
library(GenomicFeatures)
library(VariantAnnotation)
library(BiocParallel)

#Zadanie3&4
setwd("C:/Users/USER/Desktop/bioinformatyka")
bam <- BamFile("C:/Users/USER/Desktop/bioinformatyka/aligned_sample.BAM")
fa <- FaFile("C:/Users/USER/Desktop/bioinformatyka/ecoli_genome.fna")

input_bam <- "C:/Users/USER/Desktop/bioinformatyka/aligned_sample.BAM"
output_bam <- "C:/Users/USER/Desktop/bioinformatyka/sorted_aligned_sample.BAM"
sortBam(file = input_bam, destination = output_bam, overwrite = TRUE)
sorted_bam <- "C:/Users/USER/Desktop/bioinformatyka/sorted_aligned_sample.BAM.bam"

indexFa(fa)
indexBam(sorted_bam)

#Zadanie5
scanBamHeader(bam)

$targets
U00096.3 
 4641652 

$text
$text$`@HD`
[1] "VN:1.0"      "SO:unsorted"

$text$`@SQ`
[1] "SN:U00096.3" "LN:4641652" 

$text$`@PG`
[1] "ID:subread"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              
[2] "PN:subread"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              
[3] "VN:Rsubread 2.20.0"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
[4] "CL:\"subread-align\" \"-r\" \"C:\\Users\\USER\\Desktop\\bioinformatyka\\SRR31384611.fastq\" \"-o\" \"C:\\Users\\USER\\Desktop\\bioinformatyka\\aligned_sample.BAM\" \"-i\" \"C:\\Users\\USER\\Documents\\ecoli_index\" \"--type\" \"0\" \"-n\" \"10\" \"-m\" \"3\" \"-p\" \"1\" \"-M\" \"3\" \"-T\" \"1\" \"-I\" \"5\" \"--multiMapping\" \"-B\" \"1\" \"-d\" \"50\" \"-D\" \"600\" \"-S\" \"fr\" \"--trim5\" \"0\" \"--trim3\" \"0\" \"-G\" \"-1\" \"-E\" \"0\" \"-X\" \"0\" \"-Y\" \"2\" \"-P\" \"3\" "

idxstats <- idxstatsBam(sorted_bam)
print(idxstats)

eqnames seqlength  mapped unmapped
1 U00096.3   4641652 1278221        0
2        *         0       0   365453

coverage_data <- coverage(sorted_bam)
summary(coverage_data[[1]])

 Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   0.00   44.00   66.00   64.41   87.00  782.00 

#Zadanie 6   
pileup_param <- PileupParam(
    distinguish_strands = FALSE,
    distinguish_nucleotides = TRUE,
    min_base_quality = 20
)
pile <- pileup(sorted_bam, scanBamParam = ScanBamParam(), pileupParam = pileup_param)

library(dplyr)

pile_df <- pile %>%
    mutate(seqnames = as.character(seqnames)) %>%
    mutate(seqnames = ifelse(seqnames == "ABWKJP000000000.1", "NC_000913.3", seqnames))
    
#od tego momentu nic nie chciało działać i nie wiedziałam jak to zmienić

variant_candidates <- pile_df %>%
    group_by(seqnames, pos) %>%
    summarise(
        total = sum(count),
        A = sum(count[nucleotide == "A"]),
        C = sum(count[nucleotide == "C"]),
        G = sum(count[nucleotide == "G"]),
        T = sum(count[nucleotide == "T"]),
        .groups = 'drop'
    ) %>%
    mutate(
        ref = as.character(getSeq(fa, GRanges(seqnames, IRanges(pos, pos))))
    ) %>%
    rowwise() %>%
    mutate(
        alt_alleles = list(setdiff(c("A", "C", "G", "T"), ref)),
        ref_count = sum(c_across(c("A", "C", "G", "T"))[ref]),
        alt_count = sum(c_across(c("A", "C", "G", "T"))[alt_alleles])
    ) %>%
    ungroup() %>%
    
filter(alt_count >= 5) %>%
head(variant_candidates)

#Zadanie7
filtered_variants <- variant_candidates %>%
    filter(total >= 10, alt_count / total >= 0.2, alt_count >= 5)

cat("Liczba wariantów przed filtrowaniem:", nrow(variant_candidates), "\n")
cat("Liczba wariantów po filtrowaniu:", nrow(filtered_variants), "\n")

df_variants <- as.data.frame(filtered_variants)

write.csv(df_variants, "C:/Users/USER/Desktop/bioinformatyka/wyniki_wariantow.csv", row.names = FALSE)
